### PR TITLE
interfaces: add screencast-legacy for video and audio recording

### PR DESCRIPTION
**Pull request by jdstrand**
gnome-shell currently provides the org.gnome.Shell.Screencast and
org.gnome.Shell.Screenshot interfaces which is currently the only way that
the gnome-shell on wayland can perform screenshots and screencasts. See:
https://help.gnome.org/users/gnome-help/stable/screen-shot-record.html

Implement this in its own 'screenshot' implicit classic interface that is not
auto-connected.

Reference: https://forum.snapcraft.io/t/screen-recording-on-wayland/3563

== Original ==
Implement this as part of desktop-legacy for a number of reasons:
- desktop-legacy is the place for temporary or unsafe APIs until a suitable
  safe replacement is found
- screencast and screenshot would ideally be handled by a dedicated service.
  For GNOME applications, that would be the Pipewire (aka pulsevideo) as
  surfaced by portals. Not all systems will have this portal, so put in
  desktop-legacy until it is widely available
- desktop-legacy already has an expansive surface via the accessibility bus
  and input methods, which already allows access to all user input. The
  screencast and screenshot methods expand this surface to what is displayed
  on the screen, though gnome-shell will indicate that screen recording is
  taking place and play a sound when a screen shot is taken (though the sound
  can be subverted via the gsettings interface)

An argument can be made for separating this out into a dedicated, manually connected interface. IME, since the current desktop-legacy, unity7, x11 and home interfaces are auto-connected and their access is much worse than what screenshot and screencast add, adding this to `desktop-legacy` seems 'reasonable' (with the understanding that this API is temporary until the portal is available). Indeed, that is specifically what this interface is meant to address: iffy accesses that we allow until such point that applications plugging only `desktop` can be expected to run correctly.

If people (hello @niemeyer!) want this split out into a separate interface, then we would have to come up with a good interface name (this DBus API is only provided by gnome-shell).

Reference: https://forum.snapcraft.io/t/screen-recording-on-wayland/3563

### CODE REVIEW

**Comment by jdstrand** [comment]
I tested this interface using the 'peek' deb (the peek snap did not work correctly for reasons unrelated to this PR). Running 'peek' under confinement I found many GVFS policy violations since it appears to try to use it for file operations, though these violations did not affect the application and peek fell back gracefully and could save files fine. I could make these violations go away with:

```
# info leaks
dbus (send)
    bus=session
    interface=org.gtk.vfs.MountTracker
    member=ListMounts
    peer=(label=unconfined),
dbus (send)
    bus=session
    path=/org/gtk/vfs/Daemon
    interface=org.gtk.vfs.Daemon
    member=ListMonitorImplementations
    peer=(label=unconfined),
dbus (send)
    bus=session
    path=/org/gtk/Private/RemoteVolumeMonitor
    interface=org.gtk.Private.RemoteVolumeMonitor
    member={IsSupported,List}
    peer=(label=unconfined),
dbus (send)
    bus=session
    path=/org/gtk/vfs/mount/**
    interface=org.gtk.vfs.Mount
    member={Enumerate,QueryInfo}
    peer=(label=unconfined),
dbus (receive)
    bus=session
    path=/org/gtk/vfs/client/enumerator/**
    interface=org.gtk.vfs.Enumerator
    peer=(label=unconfined),

# confinement escape
dbus (send)
    bus=session
    path=/org/gtk/vfs/metadata
    interface=org.gtk.vfs.Metadata
    member=Remove
    peer=(label=unconfined),
dbus (send)
    bus=session
    path=/org/gtk/vfs/Daemon
    interface=org.gtk.vfs.Daemon
    member=GetConnection  # this only gives us the unix address, below
    peer=(label=unconfined),
unix (connect, receive, send)
    type=stream
    peer=(addr="@/dbus-vfs-daemon/socket-*"),
```

I considered put the above into desktop-legacy (and unity7), but opted against it at this time. The 'org.gtk.vfs.Metadata.Remove' allows changing metadata in ~/.local/share/gvfs-metadata. I didn't verify through code inspection, but based on my [reading](https://wiki.gnome.org/Projects/gvfs/doc) it looks like the socket access would allow copy/move via the gvfsd daemon, thus allowing sandbox escape.

See: https://git.gnome.org/browse/gvfs/tree/common/org.gtk.vfs.xml

**Comment by bboozzoo** [comment]
Any idea if kwin guys are brewing a similar solution?

**Comment by mvo5** [comment]
Setting to blocked because there is an open question for Gustavo in here.

**Comment by jdstrand** [comment]
@niemeyer - "For the case at hand, I suggest "screenshot" as the interface name. That covers both pictures and long term recording, and is a term people can clearly associate with and tell that the application can see what they can see."

That's fine. We can also make this one manually connected by default if we do it this way. I'll update.

**Comment by bboozzoo** [comment]
FWIW there's some upcoming stuff on the portals site which @jhenstridge is probably already aware of
https://github.com/flatpak/xdg-desktop-portal/blob/master/data/org.freedesktop.impl.portal.ScreenCast.xml
https://github.com/flatpak/xdg-desktop-portal/blob/master/data/org.freedesktop.impl.portal.Screenshot.xml

**Comment by jhenstridge** [comment]
For what it is worth, I have screenshots working through xdg-desktop-portal with snapd master.  Using that API, the user is prompted to share the screenshot with the application, and it is then it is shared with the app through the document portal.

I'd also be cautious about exposing the direct gnome-shell API though: it is essentially a "write a PNG file anywhere" API, without regard for the requesting application's confinement:

https://gitlab.gnome.org/GNOME/gnome-shell/blob/master/data/org.gnome.Shell.Screenshot.xml#L14

I know we've got a bunch of other insecure APIs in `desktop-legacy` (e.g. at-spi effectively lets you remote control any other application in the session), so I'm not sure where this falls on the spectrum.

**Comment by jhenstridge** [comment]
For screencasting in particular, we don't currently have everything wired up with xdg-desktop-portal on Ubuntu yet: it relies on PipeWire, and Ubuntu 18.04 has xdg-desktop-portal compiled with the PipeWire support disabled.

From the look of the gnome-shell API, there are similar concerns about confined apps being able to place files in arbitrary locations that they can not access themselves.

**Comment by jdstrand** [comment]
@jhenstridge - nice to hear that portals is doing this the screenshotting. Note that this PR has changed and it is *not* part of the desktop-legacy interface any more and instead is in a separate, manually connected interface. This will allow your portals to just start working when using 'desktop', nothing is changed for people using 'desktop-legacy' and only applications using the 'screenshot' interface will get this access when manually connected (or a snap declaration is granted).

I would be 'ok' if this PR was closed on the basis that applications should 'just use portals', but that will mean that on systems that don't have your portals work installed, these snaps won't work.

File: interfaces/builtin/screenshot.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+const screenshotSummary = `allows privileged access to desktop screenshot, screencast and recording with saving result to arbitrary locations`
```
Comment (niemeyer) [inline interfaces/builtin/screenshot.go]: "allows screen recording and audio recording, and also writing to arbitrary filesystem paths"

**Comment by jdstrand** [comment]
@niemeyer, et al - sorry for the delay on this. I've responded to all the feedback in this PR (ie, renamed as screencast-legacy).

**Pull request merged**